src/View/SSTemplateEngine.php

```diff
@@ -0,0 +1,468 @@
+
+use InvalidArgumentException;
+use Psr\SimpleCache\CacheInterface;
+use SilverStripe\Control\Director;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Flushable;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Core\Kernel;
+use SilverStripe\Core\Path;
+use SilverStripe\ORM\FieldType\DBHTMLText;
+use SilverStripe\Security\Permission;
+use SilverStripe\View\Exception\MissingTemplateException;
+
+/**
+ * Parses template files with an *.ss file extension, or strings representing templates in that format.
+ *
+ * In addition to a full template in the templates/ folder, a template in
+ * templates/Content or templates/Layout will be rendered into `$Content` and
+ * `$Layout`, respectively.
+ *
+ * A single template can be parsed by multiple nested SSTemplateEngine instances
+ * through `$Layout`/`$Content` placeholders, as well as `<% include MyTemplateFile %>` template commands.
+ *
+ * <b>Caching</b>
+ *
+ * Compiled templates are cached, usually on the filesystem.
+ * If you put ?flush=1 on your URL, it will force the template to be recompiled.
+ *
+ */
+class SSTemplateEngine implements TemplateEngine, Flushable
+{
+    use Injectable;
+    use Configurable;
+
+    /**
+     * Default prepended cache key for partial caching
+     */
+    private static string $global_key = '$CurrentReadingMode, $CurrentUser.ID';
+
+    /**
+     * List of models being processed
+     */
+    protected static array $topLevel = [];
+
+    /**
+     * @internal
+     */
+    private static bool $template_cache_flushed = false;
+
+    /**
+     * @internal
+     */
+    private static bool $cacheblock_cache_flushed = false;
+
+    private ?CacheInterface $partialCacheStore = null;
+
+    private ?TemplateParser $parser = null;
+
+    /**
+     * A template or pool of candidate templates to choose from.
+     */
+    private string|array $templateCandidates = [];
+
+    /**
+     * Absolute path to chosen template file which will be used in the call to render()
+     */
+    private ?string $chosen = null;
+
+    /**
+     * Templates to use when looking up 'Layout' or 'Content'
+     */
+    private array $subTemplates = [];
+
+    public function __construct(string|array $templateCandidates = [])
+    {
+        if (!empty($templateCandidates)) {
+            $this->setTemplate($templateCandidates);
+        }
+    }
+
+    /**
+     * Execute the given template, passing it the given data.
+     * Used by the <% include %> template tag to process included templates.
+     *
+     * @param array $overlay Associative array of fields (e.g. args into an include template) to inject into the
+     * template as properties. These override properties and methods with the same name from $data and from global
+     * template providers.
+     */
+    public static function execute_template(array|string $template, ViewLayerData $data, array $overlay = [], ?SSViewer_Scope $scope = null): string
```

Comment:
This method only exists because the template parser wants to call it for `<% include sometemplate %>` and I couldn't be bothered refactoring how that all works out.